libraries/Microsoft.Bot.Connector/Authentication/ServiceClientCredentialsFactory.cs

```diff
@@ -27,6 +27,14 @@ public abstract class ServiceClientCredentialsFactory
         /// </para></remarks>
         public abstract Task<bool> IsValidAppIdAsync(string appId, CancellationToken cancellationToken);
 
+        /// <summary>
+        /// Get App password for given App id.
+        /// </summary>
+        /// <param name="appId">The App id.</param>
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <returns>The App password.</returns>
+        public abstract Task<string> GetAppPasswordAsync(string appId, CancellationToken cancellationToken);
```

Comment:
As a general note, we can never add abstract methods to already shipped public classes. If a bot framework user inherited from this class and we ship a new abstract methods, their code would stop building because of this backward compatibility violation. We don't break backward compat unless on well planned major releases which have happened only once in the last few years.

Reply:
For reference, this class was already shipped in 4.12 so it would enter the terrain where we have to be backward compatible. https://github.com/microsoft/botbuilder-dotnet/blob/4.12/libraries/Microsoft.Bot.Connector/Authentication/ServiceClientCredentialsFactory.cs

Reply:
Added implementation.